Fix regression introduced by fix for Symfony 3, check for undefined schemes key

tobias-93 opened the pull request:
Fix for #287, problem introduced by #254

GuilhemN commented:
Any Idea why the tests are failing?

tobias-93 commented:
@GuilhemN I think the tests are a little more strict than browsers are. I just added another check which should not trigger the test to fail.

tobias-93 commented:
Ok, I'll have to look into them, as they still fail

willdurand commented:
I see a lot of:

```
TypeError: 'undefined' is not an object (evaluating 'route.schemes.length') 'undefined' is not an object (evaluating 'route.schemes.length') TypeError: 'undefined' is not an object (evaluating 'route.schemes.length')
```

maguedon commented:
I think it is because the property "schemes" may not exist in some cases. You should check if it exists first. Why did you remove the condition `goog.object.containsKey(route, "schemes")` ? I suppose it checks if "schemes" exists. Does it work if you add it at the beginning of your condition like this :

``` js
else if (goog.object.containsKey(route, "schemes") && route.schemes.length > 0 && typeof route.schemes[0] !== "undefined" && this.getScheme() != route.schemes[0]) {
```

tobias-93 commented:
I forgot to add the new object keys to the JS tests, so it should be fixed now

tobias-93 commented:
The tests now work again. @maguedon that would also fix it, but we can assume this key will always exist since the Symfony command always adds it to the object.

willdurand commented:
> The tests now work again. @maguedon that would also fix it, but we can assume this key will always exist since the Symfony command always adds it to the object.

Is that 100% sure? For BC purpose I would rather go for a test and not assume it exists..

tobias-93 commented:
@willdurand It should be. I added the check anyway, just to be sure.

GuilhemN commented on Resources/js/router.js:
```diff
@@ -241,7 +241,7 @@ fos.Router.prototype.generate = function(name, opt_params, absolute) {
     if (goog.object.containsKey(route.requirements, "_scheme") && this.getScheme() != route.requirements["_scheme"]) {
         url = route.requirements["_scheme"] + "://" + (host || this.getHost()) + url;
-    } else if (goog.object.containsKey(route, "schemes") && this.getScheme() != route.schemes[0]) {
+    } else if (goog.object.containsKey(route, "schemes") && route.schemes.length > 0 && typeof route.schemes[0] !== "undefined" && this.getScheme() != route.schemes[0]) {
```
I don't think we need both checks :) (`.length` and `!== undefined`)

tobias-93 replied:
I agree, doesn't do any harm though

GuilhemN commented:
Is this ok to merge @willdurand ?

willdurand commented:
It would be better not to update the whole test suite to ensure BC but otherwise sure :+1:

tobias-93 commented:
@willdurand I have removed the updates to the test suite, so just the new tests remain.
@GuilhemN I have removed the duplicate check.
Is this ready to be merged now?

willdurand commented:
Looks good to me!

willdurand commented:
Looks good to me!
